### ISSUE TITLE
False negative with a constrained generic type in the presence of `Any`

### ISSUE THREAD

**Issue by russelldavis** [python/mypy#9186]
```python
from typing import Any, AnyStr
def foo(arg1: AnyStr, arg2: AnyStr) -> None: ...
bar: Any = ""
foo(1, "") # Error correctly reported
foo(b"", bar) # No error (correct)
foo(1, bar) # No error (false negative)
```

In the presence of `Any`, I would expect mypy to allow any of the values of the constraint (in this case, `str` or `bytes`), but not any arbitrary type.

Playground link: https://mypy-play.net/?mypy=latest&python=3.8&flags=strict&gist=6970b49fa90d7767b0ef26c9f977e182

**Comment by ilevkivskyi** [comment]
Hm, interesting, likely this is caused by the fact that `Any` is a subtype of every type and mypy infers `Any` as the value for `AnyStr` (as a join of `int` and `Any`), and them when checking whether it is a valid value for `AnyStr` it concludes that it is.

It may actually be not easy to fix, since we want `foo(bar, bar)` to work. A while ago I proposed to make `join(int, Any)` to return `int`, but people were not happy about this since it may cause what some may see as a false positives in other places.
